docs: import export for project/env (#3369)
Unleash/unleash: website/docs/reference/deploy/environment-import-export.mdx

## Export

You can export features either from a project search or the global feature search. Use the search functionality to narrow the results to the list of features you want to export and use the export button to select environment. All features included in your search results will be included in the export.

<Figure caption="Search-driven export screen with WYSIWYG mode."  img="/img/export.png"/>

## Import

Import is a 3 stage process designed to be efficient and error-resistant.

### Import stages

* **upload** - you can upload previously exported JSON file or copy-paste export data from the exported JSON file into the code editor
* **validation** - you will get feedback on any errors or warnings before you do the actual import. This makes sure your feature flags configurations. You will not be able to finish the import if you have errors. Warnings don't stop you from importing.
* **import** - the actual import that creates a new configuration in the target environment or creates a [change request](../change-requests.md) when the environment has change requests enabled

![Three-staged import process](/img/import.png)

> ### Import rejection

Reviewer: Maybe we can rephrase this header and section to be "import requirements" instead of import rejection?

```suggestion
### Import requirements
```

> An import will be rejected when there's a possibility of data corruption.

Reviewer: What does this mean? How do we identify data corruption? Or are we talking about any potential issues that may stop the import (as listed in the subsections)?

Like this?

```suggestion
Unleash will reject an import if it discovers conflicts between the data to be imported and what already exists on the Unleash instance. The import tool will tell you about why an import is rejected in these cases. 

The following sections describe requirements that must be met for Unleash not to stop the import job.
```

Author: YEs

#### Context fields

> The exported and imported context fields have the same name but the target environment doesn't support some of the source environment context field values. If the context fields don't exist in the target environment they will be created.

Reviewer: I still don't understand this. What are the imported context fields and what are the exported context fields? Is this about conflicts between context fields that you are trying to import and context fields that already exist in the Unleash instance that you're trying to import them into?

I think there's three cases here, right? (Note: the "new" context field is the imported one, the "old" context field is the one that already exists in the instance)
1. You import a context field that doesn't exist in the target instance. All good.
2. You import a context field that already exists in the target instance. The new and the old context fields either have the exact same legal values or both accept everything. This is fine.
3. You import a context field that already exists in the target instance. There is a mismatch between the old and the new one. One of:
  i. The new context field has no restrictions on the value, but the existing one has a defined set of legal values
  ii. The old context field has no restrictions on the value, but the new field has a defined set of legal values
  iii. Both fields have legal values, but they are **not exactly the same** (even if the new custom context field's values are a strict subset of the old field's values)

Is that correct?

Author: The actual algorithms is much simpler: https://github.com/Unleash/unleash/blob/main/src/lib/features/export-import-toggles/import-context-validation.ts
Here's the test for it: https://github.com/Unleash/unleash/blob/main/src/lib/features/export-import-toggles/import-context-validation.test.ts

In essence, if you're importing legal values that don't exist in the target instance you get an error.

Reviewer: Yeah, okay, so it's more like two:
1. The imported field doesn't exist already. All good.
2. If the imported context field has legal values, then the existing context field must also have at least those values defined. Any other values in the existing field is also okay.

Reviewer: ```suggestion
When you import a custom context field **with legal values defined**:

If a custom context field with the same name already exists in the target instance, then the pre-existing context field must have **at least** those legal values defined. In other words, the imported context field legal values must be a subset of the already existing context field's legal values.

When you import custom context fields without legal values or custom context fields that don't already exist in the target instance, then there are no requirements. 

Custom context fields that don't already exist in the target instance will be created upon import.
```

#### Segments

> The target Unleash instance doesn't have a matching segment. The actual constraints in the segment can be different. The matching is done by the segment name.

Reviewer: ```suggestion
If your import has segments, then a segment with the same name must already exist in the Unleash instance you are trying to import into. Only the name must be the same: the constraints in the segment can be different.
```

#### Custom strategies

> The exported strategies include custom strategies that don't exist in the target Unleash instance. The matching is done by the strategy name.

Reviewer: ```suggestion
If your import contains custom strategies, then custom strategies with the same names must already exist in the target Unleash instance. The custom strategy definitions (including strategy parameters) that exist in the target instance do not otherwise need to match the custom strategy definitions in the instance the import came from.
```

#### Existing features

> You're trying to import a feature that already exists in a different project on the target instance.

Reviewer: ```suggestion
If any of the features you import already exist in the target Unleash instance, then they must exist within the project you're importing into. If any features you are attempting to import exist in a **different** project on the target instance, the import will be rejected.
```

> #### Pending change request
> 
> The target project and environment have a pending change request (applies to enterprise customers with change requests enabled for a given project and environment).

Reviewer: ```suggestion
#### Pending change requests

The project and environment you are importing into must **not have any pending [change requests](./change-requests.md)**. 
```

### Import warnings

> Import validation will warn you when:
> * the features you are trying to import are already archived in the target instance in any project. Those features won't be imported again.
> * the strategies you are trying to import are custom and may have different [configuration parameters](../custom-activation-strategies.md#parameters). Those strategies will be imported anyway.

Reviewer: So custom strategies always cause either a warning or an error, right? And why don't we validate custom strategies? Not worth the effort?

```suggestion
The import validation system will warn you about potential problems it finds in the import data. These warnings do not prevent you from importing the data, but you should read them carefully to ensure that Unleash does as you intend.

The following sections list things that the import tool will warn you about.

#### Archived features

The import tool will not import any features that have already been archived on the target Unleash instance. Because features are identified by their name, that means that if a feature called `feature-a` has been created and archived on the target Unleash instance, then a feature with the same name (`feature-a`) will not be imported.

If you permanently delete the archived `feature-a` from the target instance, then the new `feature-a` (in the import data) **will** be imported. 

#### Custom strategies

Unleash will verify that any custom strategies you are trying to import have already been defined on the target instance. However, it only does this verification by name. It does **not** validate that the definitions are otherwise the same or that they have the same [configuration parameters](../custom-activation-strategies.md#parameters). 

Before proceeding with the import, you should be sure that the custom strategy definitions are the same in the target instance and the imported custom strategies.
```

Author: Because you may want the custom strategy work differently in those two instances. So we leave it up to you to decide if you're ok exporting a custom strategy. Also my impression is that we may want to start using constraints instead of custom strategies. So not worth investing into it too much right now.

Author: Custom strategy gives a warning if you have it in a target instance. Custom strategy gives you an error if you don't have it in a target instance.

Reviewer: Got it, thanks. In that case, I think the suggestion still stands, but we could maybe reword or remove the last section:

```suggestion
The import validation system will warn you about potential problems it finds in the import data. These warnings do not prevent you from importing the data, but you should read them carefully to ensure that Unleash does as you intend.

The following sections list things that the import tool will warn you about.

#### Archived features

The import tool will not import any features that have already been archived on the target Unleash instance. Because features are identified by their name, that means that if a feature called `feature-a` has been created and archived on the target Unleash instance, then a feature with the same name (`feature-a`) will not be imported.

If you permanently delete the archived `feature-a` from the target instance, then the new `feature-a` (in the import data) **will** be imported. 

#### Custom strategies

Unleash will verify that any custom strategies you are trying to import have already been defined on the target instance. However, it only does this verification by name. It does **not** validate that the definitions are otherwise the same or that they have the same [configuration parameters](../custom-activation-strategies.md#parameters). 
```

### Required permissions

To import features, you will **always** require the **update feature toggles** permission.

> Additionally, depending on how the actions your import job would trigger, you may also require any of the following permissions:

Reviewer: Whoops, an extra word slipped in there

```suggestion
Additionally, depending on the actions your import job would trigger, you may also require any of the following permissions:
```

> * **Create feature toggles**: When the import would create new features

Reviewer: All other lines start with a lowercase letter, so let's be consistent.

```suggestion
* **Create feature toggles**: when the import would create new features
```

* **Update tag types**: when the import would create new tag types
* **Create context fields**: when the import would create new context fields

> * **Create activation strategies**: when the import would add (or is this "create custom activation strategies"? custom or standard?) activation strategies to a feature and change requests are disabled

Reviewer: I think you missed the question I added for you on this line 😅  Could you clarify whether this is for creating custom strategies or adding a strategy to a feature?

Reviewer: Okay. That's ... pretty confusing to me. But okay, I'll make a suggestion for it.

* **Delete activation strategies**: when import would remove existing activation strategies from a feature and change requests are disabled
* **Update variants**: when the import would update variants and change requests are disabled

### Import and change requests

> When the target project and environment have change requests enabled the import will not be applied instantly. Instead, it will
> create a change request, send it to review and await for the approval.

Reviewer: Something like this maybe? 

```suggestion 
If change requests are enabled for the target project and environment, the import will not be fully applied. Any new features will be created, but all feature configuration will be added to a new change request.
```

If change requests are enabled, any permissions for  **Create activation strategies**, **Delete activation strategies** and **Update variants** are not required.

## Environment Import & Export vs Old Import & Export

The new Environment Import & Export differs from the old [Import & Export](import-export.md).
The most important difference is a strong focus on a single project and environment in the target environment.

> Also, the new Import & Export has a UI and extra safety mechanisms preventing corrupted data imports.

Reviewer: I don't have the time right now, but I'll get back to this later. I think it could do with some fleshing out.

Reviewer: Maybe something like this?

```suggestion
## Environment import/export vs the [instance import/export API](import-export.md)

Environment import/export has some similarities to the [instance import/export API](import-export.md), but they serve different purposes. 

The instance import/export API was designed to export all feature toggles (optionally with strategies and projects) from one Unleash instance to another. When it was developed, Unleash had much fewer features than it does now. As such, the API lacks support for some of the more recent features in Unleash.

On the other hand, the environment import/export feature was designed to export a selection of features based on search criteria. It can only export data from a single environment and only import it to a single environment. It also only supports importing into a single project (although it can export features from multiple projects).

Further, the environment import/export comes with a much more stringent validation and will attempt to stop any corrupted data imports.
```

#### Pending change requests

The project and environment you are importing into must **not have any pending [change requests](../change-requests.md)**.

### Import warnings

Import validation will warn you when:
* the features you are trying to import are already archived in the target instance in any project. Those features won't be imported again.
* the strategies you are trying to import are custom and may have different [configuration parameters](../custom-activation-strategies.md#parameters). Those strategies will be imported anyway.

### Required permissions

To import features, you will **always** require the **update feature toggles** permission.
Additionally, depending on the actions your import job would trigger, you may also require any of the following permissions:
* **Create feature toggles**: when the import would create new features
* **Update tag types**: when the import would create new tag types
* **Create context fields**: when the import would create new context fields

> * **Create activation strategies**: when the import would add an activation strategies to a feature and change requests are disabled

Reviewer: ```suggestion
* **Create activation strategies**: when the import would add activation strategies to a feature and change requests are disabled
```

On the project-level these  items are exported:

* [the feature itself](../feature-toggles.mdx)
* [feature tags](../tags.md)

On the environment-level, these items are exported for the chosen environment:

* [activation strategies](../activation-strategies.md) including [constraints](../strategy-constraints.md) and references to [segments](../segments.mdx)
* [variants](../feature-toggle-variants.md)
* enabled/disabled

Additionally, these global configuration items are exported:
* [custom context fields](../unleash-context.md#custom-context-fields)
* [feature tag types](../tags.md#tag-types)

Importantly, while references to [segments](../segments.mdx)) are exported, the segments themselves are **not** exported. Consult the [import rejection](#import-rejection) section for more information.

## Export

You can export features either from a project search or the global feature search. Use the search functionality to narrow the results to the list of features you want to export and use the export button to select environment. All features included in your search results will be included in the export.

> <Figure caption="Search-driven export screen with WYSIWYG mode."  img="/img/export.png"/>

Reviewer: ```suggestion
<Figure caption="Feature toggle lists can be filtered using the search bar. Once filtered, you can use the "export current selection" button to export the configuration for those features."  img="/img/export.png"/>
```
